### PR TITLE
add two datasets: `PolBlogs` and `EmailEUCore`

### PR DESCRIPTION
**Pull request by AlexDuvalinho**
Addition of two useful existing datasets, found on external sources.

### CODE REVIEW

File: torch_geometric/datasets/__init__.py [hunk 1]
```diff
@@ -138,6 +140,8 @@
     'BAShapes',
     'MalNetTiny',
     'OMDB',
+    'Polblogs',
+    'Email_eu_core',
```
Comment (rusty1s) [inline torch_geometric/datasets/__init__.py:144]:
```suggestion
    'EmailEUCore',
```
I think we do not want to use snake-case in class names. Prefer `EmailEUCore`.

File: torch_geometric/datasets/email_eu_core.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import os
+
+import networkx as nx
+import torch
+from torch_geometric.data import InMemoryDataset, download_url
+from torch_geometric.utils.convert import from_networkx
+
+
+class Email_eu_core(InMemoryDataset):
+    """ This dataset is an e-mail communication network from a large European
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:11]:
```suggestion
    r"""This dataset is an e-mail communication network from a large European 
```

File: torch_geometric/datasets/email_eu_core.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+import networkx as nx
+import torch
+from torch_geometric.data import InMemoryDataset, download_url
+from torch_geometric.utils.convert import from_networkx
+
+
+class Email_eu_core(InMemoryDataset):
+    """ This dataset is an e-mail communication network from a large European 
+    research institution, taken from the "Local Higher-order Graph Clustering."
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:12]:
```suggestion
    research institution, taken from the `"Local Higher-order Graph Clustering" <https://www-cs.stanford.edu/~jure/pubs/mappr-kdd17.pdf>`_ paper. 
```

File: torch_geometric/datasets/email_eu_core.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    """ This dataset is an e-mail communication network from a large European 
+    research institution, taken from the "Local Higher-order Graph Clustering." 
+    paper <https://www-cs.stanford.edu/~jure/pubs/mappr-kdd17.pdf>.  
+    Vertices indicate members of the institution and an edge between a pair 
+    of members indicates that they exchanged at least one email. 
+    Vertex labels indicate membership to one of the 42 departments.
+
+    Args:
+        root (string): Root directory where the dataset should be saved.
+        split (string): If :obj:`"train"`, loads the training dataset.
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:20]: It seems like the `split` argument is not used.

File: torch_geometric/datasets/email_eu_core.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            If :obj:`"test"`, loads the test dataset. (default: :obj:`"train"`)
+        transform (callable, optional): A function/transform that takes in an
+            :obj:`torch_geometric.data.Data` object and returns a transformed
+            version. The data object will be transformed before every access.
+            (default: :obj:`None`)
+        pre_transform (callable, optional): A function/transform that takes in
+            an :obj:`torch_geometric.data.Data` object and returns a
+            transformed version. The data object will be transformed before
+            being saved to disk. (default: :obj:`None`)
+        pre_filter (callable, optional): A function that takes in an
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:31]: As the dataset only has a single example, I suggest to remove `pre_filter` as well.

File: torch_geometric/datasets/email_eu_core.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            :obj:`torch_geometric.data.Data` object and returns a boolean
+            value, indicating whether the data object should be included in the
+            final dataset. (default: :obj:`None`)
+    """
+
+    url = 'https://snap.stanford.edu/data/email-Eu-core.txt.gz'
+    url2 = 'https://snap.stanford.edu/data/email-Eu-core-department-labels.txt.gz'
+
+    def __init__(self, root, transform=None, pre_transform=None, pre_filter=None):
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:40]:
```suggestion
from typing import Optional, Callable

    def __init__(self, root: str, transform: Optional[Callable]=None, pre_transform: Optional[Callable]=None):
```

File: torch_geometric/datasets/email_eu_core.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        super().__init__(root=root, transform=transform,
+                         pre_transform=pre_transform, pre_filter=pre_filter)
+        self.data = torch.load(self.processed_paths[0])
+
+    @property
+    def raw_file_names(self):
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:46]:
```suggestion
    def raw_file_names(self) -> str:
```

File: torch_geometric/datasets/email_eu_core.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        return 'email-eu.csv'
+
+    @property
+    def processed_file_names(self):
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:50]:
```suggestion
    def processed_file_names(self) -> str:
```

File: torch_geometric/datasets/email_eu_core.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        return 'data.pt'
+
+    def download(self):
+        # If already downloaded
+        if os.path.isdir(self.raw_dir) and len(os.listdir(self.raw_dir)) > 0:
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:55]: No need to check. This is already handled internally.

File: torch_geometric/datasets/email_eu_core.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+           return
+        # Download to `self.raw_dir`.
+        for url in [self.url, self.url2]:
+            path = download_url(url, self.raw_dir)
+            path = os.path.abspath(path)
+            with gzip.open(path, 'r') as r:
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:61]: You should be able to make use of `torch_geometric.data.extract_gz` for this.

File: torch_geometric/datasets/email_eu_core.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+                with open(os.path.join(self.raw_dir, '.'.join(path.split('.')[:-1])), 'wb') as w:
+                    w.write(r.read())
+            os.unlink(path)
+
+    def process(self):
+        # Read data into huge `Data` list.
+        dirname = os.path.join(os.path.dirname(os.path.dirname(
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:68]: `dirname = self.raw_dir` should work as well.

File: torch_geometric/datasets/email_eu_core.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            #num_unique_node_labels = max(node_labels) + 1
+        except IOError:
+            print('No node labels')
+
+        # Convert networkx graph to pyg Dataset
+        data = from_networkx(G)
+
+        data.num_classes = max(data.y).item() + 1
+        data.x = torch.ones(data.num_nodes, 9)
+        data.x = torch.cat((data.x, data.y.unsqueeze(1)), dim=1)
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:96]: Why do we combine features with labels?

File: torch_geometric/datasets/email_eu_core.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+        if self.pre_filter is not None:
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:98]: Remove

File: torch_geometric/datasets/email_eu_core.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            final dataset. (default: :obj:`None`)
+    """
+
+    url = 'https://snap.stanford.edu/data/email-Eu-core.txt.gz'
+    url2 = 'https://snap.stanford.edu/data/email-Eu-core-department-labels.txt.gz'
+
+    def __init__(self, root, transform=None, pre_transform=None, pre_filter=None):
+        super().__init__(root=root, transform=transform,
+                         pre_transform=pre_transform, pre_filter=pre_filter)
+        self.data = torch.load(self.processed_paths[0])
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:43]:
```suggestion
        self.data, self.slices = torch.load(self.processed_paths[0])
```

File: torch_geometric/datasets/email_eu_core.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+        data.x = torch.ones(data.num_nodes, 9)
+        data.x = torch.cat((data.x, data.y.unsqueeze(1)), dim=1)
+
+        if self.pre_filter is not None:
+            data = self.pre_filter(data)
+
+        if self.pre_transform is not None:
+            data = self.pre_transform(data)
+
+        torch.save(data, self.processed_paths[0])
```
Comment (rusty1s) [inline torch_geometric/datasets/email_eu_core.py:104]:
```suggestion
        torch.save(self.collate([data]), self.processed_paths[0])
```

File: torch_geometric/datasets/polblogs.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+import os
+
+import pandas as pd
+import torch
+from torch_geometric.data import Data, InMemoryDataset, download_url
+from torch_geometric.utils import to_undirected
+
+
+class Polblogs(InMemoryDataset):
+    """ The Political Blogs dataset (Polblogs) was introduced
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:11]:
```suggestion
    r"""The Political Blogs dataset (Polblogs) was introduced 
```

File: torch_geometric/datasets/polblogs.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+    Polblogs is a graph with 1,490 vertices (representing political
+    blogs) and 19,025 edges (links between blogs). These links are
+    automatically extracted from a crawl of the front page of the blog. 
+    Each vertex receives a label indicating the political leaning of the blog: 
+    liberal or conservative. Edges are undirected. 
+
+    Args:
+        root (string): Root directory where the dataset should be saved.
+        split (string): If :obj:`"train"`, loads the training dataset.
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:24]: see above.

File: torch_geometric/datasets/polblogs.py [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+            If :obj:`"test"`, loads the test dataset. (default: :obj:`"train"`)
+        transform (callable, optional): A function/transform that takes in an
+            :obj:`torch_geometric.data.Data` object and returns a transformed
+            version. The data object will be transformed before every access.
+            (default: :obj:`None`)
+        pre_transform (callable, optional): A function/transform that takes in
+            an :obj:`torch_geometric.data.Data` object and returns a
+            transformed version. The data object will be transformed before
+            being saved to disk. (default: :obj:`None`)
+        pre_filter (callable, optional): A function that takes in an
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:35]: see above.

File: torch_geometric/datasets/polblogs.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+            :obj:`torch_geometric.data.Data` object and returns a boolean
+            value, indicating whether the data object should be included in the
+            final dataset. (default: :obj:`None`)
+    """
+
+    url = 'https://netset.telecom-paris.fr/datasets/polblogs.tar.gz'
+
+    def __init__(self, root=None, transform=None, pre_transform=None, pre_filter=None):
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:43]:
```suggestion
    def __init__(self, root: str=None, transform: Optional[Callable]=None, pre_transform: Optional[Callable]=None):
```

File: torch_geometric/datasets/polblogs.py [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+        super().__init__(root=root, transform=transform,
+                         pre_transform=pre_transform, pre_filter=pre_filter)
+        self.data = torch.load(self.processed_paths[0])
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:46]:
```suggestion
        self.data, self.slices = torch.load(self.processed_paths[0])
```

File: torch_geometric/datasets/polblogs.py [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+    @property
+    def raw_file_names(self):
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:49]:
```suggestion
    def raw_file_names(self) -> str:
```

File: torch_geometric/datasets/polblogs.py [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+        return 'polblogs.csv'
+
+    @ property
+    def processed_file_names(self):
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:53]:
```suggestion
    def processed_file_names(self) -> str:
```

File: torch_geometric/datasets/polblogs.py [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+        data.x = torch.ones(data.num_nodes, 9)
+        data.x = torch.cat((data.x, data.y.unsqueeze(1)), dim=1)
+
+        if self.pre_filter is not None:
+            data = self.pre_filter(data)
+
+        if self.pre_transform is not None:
+            data = self.pre_transform(data)
+
+        torch.save(data, self.processed_paths[0])
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:93]:
```suggestion
        torch.save(self.collate([data]), self.processed_paths[0])
```

File: torch_geometric/datasets/polblogs.py [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+    @property
+    def raw_file_names(self):
+        return 'polblogs.csv'
+
+    @ property
+    def processed_file_names(self):
+        return 'data.pt'
+
+    def download(self):
```
Comment (rusty1s) [inline torch_geometric/datasets/polblogs.py:56]: Please see related comments from above.

**Review by rusty1s — COMMENTED** [review]
This is great. Thank you! There is only minor clean-up to be done (adding type hints, fixing doc links). For downloading and extracting, I suggest to make use of `extract_gz`. Note that there is no need to check if downloading should be skipped, as this is already handled internally.

File: torch_geometric/datasets/polblogs.py (hunk 23 above)
Comment (AlexDuvalinho) [reply torch_geometric/datasets/polblogs.py:75]: I could apply these changes to both datasets. However, as a consequence, data.num_nodes and data.num_classes will each become a list with a single element, which is less convenient to process than an integer directly... Plus, self.slices is not very interesting there since there is a unique graph. 
If for consistency reasons, you still would like these changes to be made, notify it to me and I will change them :)

Comment (rusty1s) [reply torch_geometric/datasets/polblogs.py:75]: It's mostly for consistency. I think both `dataset[0].num_nodes` and `dataset[0].num_classes` should be of type integer, aren't they? Note that `data.num_classes` is probably redundant as this is already a property of the dataset.

**Review by rusty1s — APPROVED** [review]

**Comment by rusty1s** [comment]
Thank you!
